go-controller/pkg/ovn/controller/apbroute/external_controller.go

```diff
@@ -40,12 +40,16 @@ func (g gatewayInfoList) Insert(items ...*gatewayInfo) (gatewayInfoList, sets.Se
 	ret := append(gatewayInfoList{}, g...)
 	duplicated := sets.New[string]()
 	for _, item := range items {
+		cp := &gatewayInfo{bfdEnabled: item.bfdEnabled, gws: sets.Set[string]{}}
 		for _, ip := range item.gws.UnsortedList() {
 			if ret.HasIP(ip) {
 				duplicated = duplicated.Insert(ip)
 				continue
 			}
-			ret = append(ret, item)
+			cp.gws.Insert(ip)
+		}
+		if len(cp.gws) > 0 {
+			ret = append(ret, cp)
```

Comment:
> In regards to `ret.HasIP(ip)`, does it matter if the IP is contained in a gateway info that does not have the same value for `bfdEnabled` than the one `item` has? This does not seem to be in conformance with the `func (g gatewayInfo) Equal(g2 gatewayInfo) bool` implementation below which does compare `bfdEnabled`.

I'm not sure I follow you: with regards of `HasIP`, it is checking specifically for the IP, not for equality. As for the gatewayInfo Equal(), it is measuring if all values are found in the second gateway. 

I see both functions as different. Maybe I'm missing something?

Reply:
> Could we add a bit of gotext documentation to this method that covers its behavior regarding filtering duplicated IPs?

Sure, I will do :smile:

Reply:
let's say I have

```
a := gatewayInfoList{
  {
    bfdEnabled: true,
    gws: {"1.1.1.1"},
  },
  {
    bfdEnabled: false,
    gws: {"1.1.1.1"},
  }
}

b := gatewayInfoList{
  {
    bfdEnabled: true,
    gws: {"1.1.1.1"},
  },
}

c := gatewayInfoList{
  {
    bfdEnabled: false,
    gws: {"1.1.1.1"},
  },
}
```

The result of `b.Insert(c...)` is equal to `b` whereas one might expect it to be equal `a`. And it also means that it is not a commutative operation as `c.Insert(b...)` would lead to `c`.

I just want to make sure this is ok and properly documented because it does look weird.

Reply:
I understand but the impression here is that instead of avoiding it, you are ignoring it. Avoiding would be for example in HasIP, if `bfdEnabled` was different for both IPs, to return an error or panic if it is completely unexpected.

Or maybe it is being avoided somewhere else that I a missing. Not completely familiar with the overall code here. Here I am just looking at gatewayInfo/List.

Reply:
> I understand but the impression here is that instead of avoiding it, you are ignoring it. Avoiding would be for example in HasIP, if `bfdEnabled` was different for both IPs, to return an error or panic if it is completely unexpected.
> 
> Or maybe it is being avoided somewhere else that I a missing. Not completely familiar with the overall code here. Here I am just looking at gatewayInfo/List.

Honestly, I was avoiding it because I didn't expect it to be an issue, but you make a strong argument: attempting to add a duplicated IP with the different BFD states (enabled/disabled) should in fact throw an error and stop processing the policy. 

If I don't do this, the user might not be aware that there are different BFD states for the same IP, and the controller should not accept this inconsistency because it leads to a non-deterministic outcome: whichever policy is processed first will get its BFD status used.

On the other hand, adding a duplicated IP with the same BFD state should not be an issue and in fact it might even be expected. In this case I log a warning and that's good enough.

Thanks for pointing it out!